Feature refactor into mvvm

kieranwoodrow opened the pull request:
Moved service calls for Launchpads and Rockets out from controllers and put them into their respective repositories

kieranwoodrow committed:
Refactored code for service call

kieranwoodrow committed:
Fixed launchscreen image size

kieranwoodrow committed:
Extra cleanup with function names

kieranwoodrow committed:
Last cleanup for function name

bgouws commented on spaceR/spaceR/ViewModels/LaunchpadViewModel.swift:
```diff
@@ -9,26 +9,44 @@ import Foundation
     }
     
-    func setAllLaunchpads(launchpads: [Launchpads]) {
+    func set(launchpads: [Launchpads]) {
```
Don't think you need a function for this you can set the variable directly 😄

kieranwoodrow replied:
🤙🏼

bgouws commented on spaceR/spaceR/ViewModels/RocketViewModel.swift:
```diff
@@ -7,31 +7,53 @@
     }
     
-    func setAllRockets(rockets: [Rocket]) {
+    func set(rockets: [Rocket]) {
```
Same here, don't think you need this function you can set the variable directly 😄

kieranwoodrow replied:
🤙🏼

bgouws commented on spaceR/spaceR/ViewModels/RocketViewModel.swift:
```diff
@@ -7,31 +7,53 @@
     }
     
-    var rocketList: [Rocket] { return allRockets }
+    var rocketList: [Rocket] { return allRockets ?? [] }
```
You can put these over multiple lines: 

```
var rocketList: [Rocket] {
   return allRockets ?? []
}
```

kieranwoodrow replied:
🤙🏼

bgouws commented on spaceR/spaceR/Controllers/SignupViewController.swift:
```diff
@@ -30,7 +30,10 @@ class SignupViewController: UIViewController {
+                                     lastName: safeLastName,
+                                     email: safeEmail,
+                                     password: safePassword)
             return true
```
So this logic needs to be moved to the `ViewModel` and then handled with your `delegate`

kieranwoodrow replied:
cool

bgouws reviewed (changes requested):
Couple of enhancements 😄

kieranwoodrow committed:
cleanup for mvvm

bgouws commented on spaceR/spaceR/Controllers/LoginViewController.swift:
```diff
@@ -7,11 +7,18 @@
 import UIKit
 
-class LoginViewController: UIViewController, UITextFieldDelegate {
+class LoginViewController: UIViewController, UITextFieldDelegate, ViewModelDelegate {
```
Rather conform to these protocols in an extension underneath this class: 

`extension LoginViewController: UITextFieldDelegate { }`
`extension LoginViewController: ViewModelDelegate { } `

bgouws commented on spaceR/spaceR/Controllers/LoginViewController.swift:
```diff
@@ -7,11 +7,18 @@
+        
+    }
+    
+    func show(error: String) {
```
Shouldn't there be some functionality here 😄

bgouws commented on spaceR/spaceR/Controllers/LoginViewController.swift:
```diff
@@ -27,18 +34,15 @@ class LoginViewController: UIViewController, UITextFieldDelegate {
-            }
-        } else {
+        do {
+            try loginViewModel.validateUserCredentials(userEmail: textFieldEmailID.text,
```
Instead of having `loginViewModel.validateUserCredentials` return a `bool` why not let your `ViewModel` do the evaluation and then call on of your delegate functions based on the `ViewModels` evaluation.

bgouws commented on spaceR/spaceR/Controllers/SignupViewController.swift:
```diff
@@ -8,48 +8,34 @@
+                signUpViewModel.validateUserInput(userFirstName: userFirstName.text,
+                                          userLastName: userLastName.text,
+                                          userEmail: userEmail.text,
+                                          userPassword: userPassword.text) ? self.successfulSignup() : self.unsuccessfullSignup()
```
Same as above, instead of `signUpViewModel.validateUserInput` returning a `bool` value why not let your `ViewModel` do the evaluation and then call one of your delegate functions?

bgouws commented on spaceR/spaceR/ViewModels/LoginViewModel.swift:
```diff
@@ -10,32 +10,55 @@ import UIKit
+        }
+    }
+    
+    func validateUserEmail(userEmail: String?) throws -> Bool {
```
So no real need to mark this as `throws`

bgouws commented on spaceR/spaceR/ViewModels/LoginViewModel.swift:
```diff
@@ -10,32 +10,55 @@ import UIKit
+            self.userEmail = safeUserEmail
+        }
         do {
             try user = coreDataPersistantObject?.fetch(User.fetchRequest()) ?? []
```
This should be done in the `repository` 😄

bgouws commented on spaceR/spaceR/ViewModels/LoginViewModel.swift:
```diff
@@ -10,32 +10,55 @@ import UIKit
+            self.userPassword = safeUserPassword
+        }
         do {
             try user = coreDataPersistantObject?.fetch(User.fetchRequest()) ?? []
```
This should be done in the `repository`

bgouws commented on spaceR/spaceR/ViewModels/SignupViewModel.swift:
```diff
@@ -10,27 +10,40 @@ import UIKit
-    
-    func set(password: String) {
-        userPassword = password
+    func validateUserInput(userFirstName: String?, userLastName: String?,
```
No need for this function to be marked as `throws`

kieranwoodrow committed:
refactored more to architecture standards

kieranwoodrow committed:
more cleanup

kieranwoodrow committed:
Update LoginViewModel.swift

kieranwoodrow committed:
fixed duplicate error alert logic

kieranwoodrow committed:
Fixed login issue

Repository variable should not be defined as weak var, only var. If weak var is used, repository variable is remains nil

bgouws commented on spaceR/spaceR/Repositories/LoginRepository.swift:
```diff
@@ -0,0 +1,55 @@
+        self.user = []
+    }
+    
+    func fetchEmail(email: String, completion: @escaping(Email)) {
```
Remember for next PR would be better to combine these functions 👍

kieranwoodrow replied:
👍🏼

bgouws commented on spaceR/spaceR/Repositories/SignupRepository.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    private var user: [User]
+    private var successFulSignup: Bool
+    private let coreDataPersistantObject = (UIApplication.shared.delegate as?           AppDelegate)?.persistentContainer.viewContext
```
Spacing is off here 😄

kieranwoodrow replied:
thanks

bgouws commented on spaceR/spaceR/UIExtentions/UIErrorAlerts.swift:
```diff
@@ -31,6 +37,18 @@ enum CustomError: Error, LocalizedError {
+        case .invalidRequest:
+            return "Invalid request"
+        case .invalidUrl:
+            return "Inva;id URL"
```
Typo here.

kieranwoodrow replied:
ahh thanks

bgouws commented on spaceR/spaceR/UIExtentions/UIErrorAlerts.swift:
```diff
@@ -48,6 +66,18 @@ enum CustomError: Error, LocalizedError {
         case .unsuccessfulLaunchpadApiCall:
             return "Launchpad information has not been loaded successfully"
+        case .invalidResponse:
+            return "The response is invalid"
```
Seems like a lot of duplication here, can't we simplify this?

kieranwoodrow replied:
removed invalid request and invalid response. The other "invalid.." are used for the service call errors and the other errors are specific to the ui error

bgouws commented on spaceR/spaceR/ViewModels/SignupViewModel.swift:
```diff
@@ -10,42 +10,61 @@ import UIKit
+    private var repository: SignupRepositoryType?
     
-    private let coreDataPersistantObject = (UIApplication.shared.delegate as? AppDelegate)?.persistentContainer.viewContext
+    private let coreDataPersistantObject = (UIApplication.shared.delegate
```
ViewModel shouldn't know about this.

kieranwoodrow replied:
forgot to remove it, thanks

bgouws commented on spaceR/spaceR/ViewModels/SignupViewModel.swift:
```diff
@@ -10,42 +10,61 @@ import UIKit
+         delegate: ViewModelDelegate) {
+        self.delegate = delegate
+        self.repository = repository
+        self.userFirstName = ""
```
Creating a Model for all these variables would be neater 😄

kieranwoodrow replied:
so i should create a model for my model?😅
(if so, i will do it in the next pr)

bgouws replied:
So remember this is a `ViewModel` a `model` holds data for us, it is usually a struct.

bgouws reviewed (changes requested):
So this PR is bordering too big, please make sure that these PRs don't grow over time. Otherwise looks pretty clean -just a couple of enhancements.

kieranwoodrow committed:
cleanup

bgouws commented on spaceR/spaceR/Controllers/LaunchpadViewController.swift:
```diff
@@ -63,3 +48,16 @@ extension LaunchpadViewController: UITableViewDelegate, UITableViewDataSource {
+extension LaunchpadViewController: ViewModelDelegate {
+    
+    func reloadView() {
+        DispatchQueue.main.async {
```
Move this line to the repo 😄

kieranwoodrow replied:
where in the repo? Because if i don't have it here, i get this purple warning saying "UITableView.reloadData() must be used from main thread only" and then my App crashes when trying to load the rocket data

bgouws replied:
So URLSession pushes us to a background thread so that the service call doesn't happen on the main thread, but once the service call is finished we can move back to the main thread, so essentially why the app is having an issue is because you were trying to update UI from a background thread hence wrapping your UI update in `DispatchQueue.main.async` but ideally we don't want to litter our `ViewControllers` with these types of updates so rather than putting it here put it in your repo on completion (essentially once the service call is finished and you send back your data) that would be a good time to switch back to the main thread. (This goes for API and error 😄 )

kieranwoodrow replied:
okay, will have a look into it. Thanks

kieranwoodrow replied:
Okay, I put it in the service call because that's where the success completion is returned from. Is that correct?

kieranwoodrow replied:
* And the error has been removed as well. All is working as it should😁

bgouws commented on spaceR/spaceR/Controllers/RocketsControlller.swift:
```diff
@@ -63,3 +48,16 @@ extension RocketsController: UITableViewDelegate, UITableViewDataSource {
+extension RocketsController: ViewModelDelegate {
+    
+    func reloadView() {
+        DispatchQueue.main.async {
```
Move this line to the repo 😄

kieranwoodrow replied:
same as above

bgouws reviewed (commented):
Just some more clean up and then I'll approve 👍

kieranwoodrow committed:
Moved main.async to repo

  DispatchQueue.main.async should not be in view controller, rather implemented it by the success completion. If it is not used, the app crashes because you cannot update UI on the background thread. it gets pushed to the background thread when making the URLsession

bgouws reviewed (approved):
Looks good 😄

erikdvt commented:
Looks good to me. 👍
